### PR TITLE
[EXPLORER][SHPOLICY] "NoSimpleStartMenu" policy

### PR DESCRIPTION
**Pull request by feel-the-dz3n**
## Purpose

Implement "NoSimpleStartMenu" policy 

## Proposed changes

- :heavy_check_mark:  "NoSimpleStartMenu" policy
image

## TODO

- [x] [**Fix SHPolicy bug with backslash**](https://github.com/reactos/reactos/pull/1267)
- [x] Implement basic functions in explorer
- [x] It should hide radion button and labels, not disable.
- [x] Make it work without any vandalism

### CODE REVIEW

**Comment by yagoulas** [comment]
Check if SHRestricted should be used instead

File: base/shell/explorer/settings.cpp [hunk 1]
```diff
@@ -73,6 +74,11 @@ BOOL TaskbarSettings::Load()
         if (sr.Position > ABE_BOTTOM)
             sr.Position = ABE_BOTTOM;
     }
+    
+    /* Policies Settings */
+    cbSize = sizeof(NULL); // dwValue was NULL at the start, is it right?
+    dwRet = SHGetValueW(hkExplorerPolicies, NULL, L"NoSimpleStartMenu", NULL, &dwValue, &cbSize);
```
Comment (ThFabba) [inline base/shell/explorer/settings.cpp]: Values in the `Policies` key should be looked up using the `SHRestricted` function, not manually. You want `REST_NOSTARTPANEL` (which Wine's SHRestricted doesn't implement, but it should be straightforward to just add it to this table: https://github.com/reactos/reactos/blob/master/dll/win32/shell32/wine/shpolicy.c#L135)

**Comment by feel-the-dz3n** [comment]
> SHRestricted

:smile: I didn't know about "SHRestriced". I'm going to review everything again)

**Comment by feel-the-dz3n** [comment]
> Check if SHRestricted should be used instead

Also, I can't find "NoSimpleStartMenu" in ``shpolicy.c``. Should I add it there?

File: base/shell/explorer/settings.cpp (hunk 1 above)
Comment (feel-the-dz3n) [reply base/shell/explorer/settings.cpp]: Thank you for answer, I will update code later)

File: boot/bootdata/hivedef.inf [hunk 2]
```diff
@@ -1898,6 +1898,9 @@ HKCU,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon",,0x00000012
 HKCU,"SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon","Shell",0x00020000,"%SystemRoot%\explorer.exe"
 
 HKCU,"SOFTWARE\Microsoft\Windows\CurrentVersion\Policies",,0x00000012
+HKCU,"SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\Explorer",,0x00000012
+HKCU,"SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\Explorer","NoSimpleStartMenu",0x00010003,1
```
Comment (feel-the-dz3n) [inline boot/bootdata/hivedef.inf]: ``SHRestricted`` looks for value firstly in HKLM. There is no Modern start menu in ReactOS, so ``NoSimpleStartMenu`` in registry should always be ``1``. I added default value to "hivedef.inf", but there values only for HKCU. In which registry-file but for HKLM I have to add default value?

File: dll/win32/shell32/wine/shpolicy.c [hunk 3]
```diff
@@ -131,10 +131,17 @@ static const char strNoFileMenu[] = {"NoFileMenu"};
 static const char strNoSaveSetting[] = {"NoSaveSettings"};
 static const char strNoClose[] = {"NoClose"};
 static const char strNoRun[] = {"NoRun"};
+static const char strNoSimpleStartMenu[] = {"NoSimpleStartMenu"};
 
 /* policy data array */
 static POLICYDATA sh32_policy_table[] =
 {
+  {
```
Comment (ThFabba) [inline dll/win32/shell32/wine/shpolicy.c]: Keep this sorted by the value of the REST_xxx constant please, i.e. this goes between REST_NONLEGACYSHELLMODE and REST_STARTRUNNOHOMEPATH.

File: boot/bootdata/hivedef.inf (hunk 2 above)
Comment (ThFabba) [reply boot/bootdata/hivedef.inf]: You want hivesft.inf

File: base/shell/explorer/settings.cpp [hunk 4]
```diff
@@ -73,6 +74,9 @@ BOOL TaskbarSettings::Load()
         if (sr.Position > ABE_BOTTOM)
             sr.Position = ABE_BOTTOM;
     }
+    
+    /* Policies Settings */
+    bPolicyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL);
```
Comment (ThFabba) [inline base/shell/explorer/settings.cpp]: SHRestricted already handles caching (and will handle changes made to the registry without requiring an explorer restart), so don't save this value in explorer, just query it every time it's needed.

**Comment by feel-the-dz3n** [comment]
I will update code later. Thank you, Thomas)

File: base/shell/explorer/trayprop.cpp [hunk 5]
```diff
@@ -196,11 +196,22 @@ class CStartMenuSettingsPage : public CPropertyPageImpl<CStartMenuSettingsPage>
         HWND hwndCustomizeClassic = GetDlgItem(IDC_TASKBARPROP_STARTMENUCLASSICCUST);
         HWND hwndCustomizeModern = GetDlgItem(IDC_TASKBARPROP_STARTMENUCUST);
         HWND hwndStartBitmap = GetDlgItem(IDC_TASKBARPROP_STARTMENU_BITMAP);
+        HWND hwndModernRadioBtn = GetDlgItem(IDC_TASKBARPROP_STARTMENU);
+        BOOL policyNoSimpleStartMenu = (int)SHRestricted(REST_NOSTARTPANEL) == 1 ? TRUE : FALSE;
```
Comment (feel-the-dz3n) [inline base/shell/explorer/trayprop.cpp]: It always returns 0

Comment (learn-more) [reply base/shell/explorer/trayprop.cpp]: Don't check for `== 1`,
use either:
`BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) ? TRUE : FALSE;`
or:
`BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) != 0 ? TRUE : FALSE;`

Comment (feel-the-dz3n) [reply base/shell/explorer/trayprop.cpp]: Updated

Comment (ThFabba) [reply base/shell/explorer/trayprop.cpp]: Actually `? TRUE : FALSE` is an anti-pattern, it should be avoided. Just `SHRestricted(...) != 0` is a boolean expression -- no conditional operator needed.

File: base/shell/explorer/trayprop.cpp [hunk 6]
```diff
@@ -197,8 +197,8 @@ class CStartMenuSettingsPage : public CPropertyPageImpl<CStartMenuSettingsPage>
         HWND hwndCustomizeModern = GetDlgItem(IDC_TASKBARPROP_STARTMENUCUST);
         HWND hwndStartBitmap = GetDlgItem(IDC_TASKBARPROP_STARTMENU_BITMAP);
         HWND hwndModernRadioBtn = GetDlgItem(IDC_TASKBARPROP_STARTMENU);
-        BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) ? TRUE : FALSE;
-        
+        BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) != 0;
```
Comment (feel-the-dz3n) [inline base/shell/explorer/trayprop.cpp]: Returns 0 always. :disappointed:

File: base/shell/explorer/trayprop.cpp (hunk 5 above)
Comment (feel-the-dz3n) [reply base/shell/explorer/trayprop.cpp]: Updated

File: base/shell/explorer/trayprop.cpp (hunk 6 above)
Comment (ThFabba) [reply base/shell/explorer/trayprop.cpp]: Yep, apparently Wine's implementation is buggy...
```
kd> da @@(regstr)
011cea9c  "Software\Microsoft\Windows\Curre"
011ceabc  "ntVersion\PoliciesExplorer"
```
Clearly a missing backslash

**Comment by feel-the-dz3n** [comment]
See new image in the initial message

File: base/shell/explorer/trayprop.cpp [hunk 7]
```diff
@@ -196,11 +196,29 @@ class CStartMenuSettingsPage : public CPropertyPageImpl<CStartMenuSettingsPage>
...
+            
+            /* Hide controls related to modern menu */
+            ::ShowWindow(hwndModernRadioBtn, SW_HIDE);
+            ::ShowWindow(hwndModernText, SW_HIDE);
+            ::ShowWindow(hwndCustomizeModern, SW_HIDE);
+        }
 
         BOOL bModern = IsDlgButtonChecked(IDC_TASKBARPROP_STARTMENU);
-        ::EnableWindow(hwndCustomizeModern, bModern);
+        ::EnableWindow(hwndCustomizeModern, policyNoSimpleStartMenu ? FALSE : bModern);
```
Comment (ThFabba) [inline base/shell/explorer/trayprop.cpp]: This ternary seems unnecessary now. And perhaps just make `bModern`'s value depend on `policyNoSimpleStartMenu`?

File: dll/win32/shell32/wine/shpolicy.c [hunk 8]
```diff
@@ -856,6 +863,9 @@ DWORD WINAPI SHRestricted (RESTRICTIONS policy)
 	}
 
 	lstrcpyA(regstr, strRegistryPolicyA);
+#ifdef __REACTOS__
+    lstrcatA(regstr, "\\");
```
Comment (feel-the-dz3n) [inline dll/win32/shell32/wine/shpolicy.c]: related to https://github.com/reactos/reactos/pull/1267

File: base/shell/explorer/trayprop.cpp [hunk 9]
```diff
@@ -196,8 +196,31 @@ class CStartMenuSettingsPage : public CPropertyPageImpl<CStartMenuSettingsPage>
...
+        BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) != 0;
+        BOOL bModern = FALSE;
+
+        /* If NoSimpleStartMenu, disable ability to use Modern Start Menu */
+        if (policyNoSimpleStartMenu)
+        {
+            /* Swich to classic */
+            CheckDlgButton(IDC_TASKBARPROP_STARTMENUCLASSIC, BST_CHECKED);
+
+            /* Disable radion button */
```
Comment (MailYouLater) [inline base/shell/explorer/trayprop.cpp]: Typo? radion => radio

**Comment by HBelusca** [comment]
@feel-the-dz3n : you can update your PR (and fix the conflict too) :)

**Comment by HBelusca** [comment]
@feel-the-dz3n : in git, when you are in your branch, can you please do: `git fetch --all && git rebase upsteam/master` (if the main reactos.git remote is called "upstream" on your side, you can check this by running `git remote -vv`). When rebasing you should fix manually the conflicts. Once this is done, you can update your branch/PR using: `git push --force-with-lease`.

**Comment by feel-the-dz3n** [comment]
Now everything is OK?

File: base/shell/explorer/lang/pl-PL.rc [hunk 10]
```diff
@@ -100,10 +100,10 @@ FONT 8, "MS Shell Dlg", 0, 0, 0x1
 BEGIN
     CONTROL "", IDC_TASKBARPROP_STARTMENU_BITMAP, "Static", SS_BITMAP | SS_SUNKEN,7, 7, 183, 108
     AUTORADIOBUTTON "&Menu Start", IDC_TASKBARPROP_STARTMENU, 7, 135, 105, 10, WS_DISABLED
```
Comment (HBelusca) [inline base/shell/explorer/lang/pl-PL.rc]: Forgot to remove WS_DISABLED.

File: base/shell/explorer/lang/pl-PL.rc [hunk 11]
```diff
@@ -100,10 +100,10 @@ FONT 8, "MS Shell Dlg", 0, 0, 0x1
...
-    LTEXT "Ten styl menu pozwala na łatwy dostęp do twoich folderów, ulubionych programów i wyszukiwania.", IDC_STATIC, 20, 145, 150, 24, WS_DISABLED
+    LTEXT "Ten styl menu pozwala na łatwy dostęp do twoich folderów, ulubionych programów i wyszukiwania.", IDC_TASKBARPROP_STARTMENUMODERNTEXT, 20, 145, 150, 24, WS_DISABLED
     PUSHBUTTON "&Dostosuj...", IDC_TASKBARPROP_STARTMENUCUST, 192, 132, 53, 14, WS_DISABLED
```
Comment (HBelusca) [inline base/shell/explorer/lang/pl-PL.rc]: Forgot to remove WS_DISABLED.

File: base/shell/explorer/lang/zh-CN.rc [hunk 12]
```diff
@@ -99,10 +99,10 @@ FONT 9, "宋体", 0, 0, 0x1
 BEGIN
     CONTROL "", IDC_TASKBARPROP_STARTMENU_BITMAP, "Static", SS_BITMAP | SS_SUNKEN,7, 7, 183, 108
     AUTORADIOBUTTON "开始菜单(&S)", IDC_TASKBARPROP_STARTMENU, 7, 135, 105, 10, WS_DISABLED
```
Comment (HBelusca) [inline base/shell/explorer/lang/zh-CN.rc]: Forgot to remove WS_DISABLED.

File: base/shell/explorer/lang/zh-CN.rc [hunk 13]
```diff
@@ -99,10 +99,10 @@ FONT 9, "宋体", 0, 0, 0x1
...
-    LTEXT "这种菜单风格让您能方便地访问您的文件夹、收藏的程序和搜索。", IDC_STATIC, 20, 145, 150, 24, WS_DISABLED
+    LTEXT "这种菜单风格让您能方便地访问您的文件夹、收藏的程序和搜索。", IDC_TASKBARPROP_STARTMENUMODERNTEXT, 20, 145, 150, 24, WS_DISABLED
     PUSHBUTTON "自定义(&C)...", IDC_TASKBARPROP_STARTMENUCUST, 192, 132, 53, 14, WS_DISABLED
```
Comment (HBelusca) [inline base/shell/explorer/lang/zh-CN.rc]: Forgot to remove WS_DISABLED.

File: base/shell/explorer/trayprop.cpp [hunk 14]
```diff
@@ -196,8 +196,31 @@ class CStartMenuSettingsPage : public CPropertyPageImpl<CStartMenuSettingsPage>
...
         HWND hwndStartBitmap = GetDlgItem(IDC_TASKBARPROP_STARTMENU_BITMAP);
+        HWND hwndModernRadioBtn = GetDlgItem(IDC_TASKBARPROP_STARTMENU);
+        HWND hwndModernText = GetDlgItem(IDC_TASKBARPROP_STARTMENUMODERNTEXT);
+        BOOL policyNoSimpleStartMenu = SHRestricted(REST_NOSTARTPANEL) != 0;
+        BOOL bModern = FALSE;
+
+        /* If NoSimpleStartMenu, disable ability to use Modern Start Menu */
+        if (policyNoSimpleStartMenu)
+        {
+            /* Swich to classic */
```
Comment (HBelusca) [inline base/shell/explorer/trayprop.cpp]: "Switch"

Comment (feel-the-dz3n) [reply base/shell/explorer/trayprop.cpp]: fixed

**Comment by feel-the-dz3n** [comment]
https://github.com/reactos/reactos/pull/1711

**Pull request merged**